### ISSUE TITLE
'split'/'index' problem for utf8

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6547]
Migrated from [rt.perl.org#22375](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22375) (status was 'resolved')

Searchable as RT22375$

**Comment by p5pRT** [comment]
### From perl@geez.org

Greetings\,

I've found a really odd behavior either with split or index\,
I'm not really sure which\.  The script below will demonstrate
the problem\.  In short index is not returning the right value
when a string reference is used vs a constant string \(that is
$string vs "abcde"\)\.  It appears that the sequence that chars
where in before they were split effects the outcome\.  The chars
are used as the substring argument of split\.

Its really weird and unfortunately also blocking a real world
problem\.

thanks\,

/Daniel

I'm using Perl 5\.8\.0 on a Redhat 8 Linux&#8203;:


begin 775 split\-utf\-break\.pl\.gz
M'XL\("\#"$UCX\`\`W\-P;&ET\+75T9BUB\<F5A&#8203;:RYP;\`\#%4L%\.XS\`4/\.\.O>\*252J5"
M1\)$J1\+>PA^6P%T\[\<6\!2Y\[0NQUK4C^X52K;C3/^'87^J?\\\.PTI>U>N\!'Y$\(\_'
M\,\_\,F&#8203;:1VGE7?I6\)FT1\*?A="XJCU\!1?CF\,;YZ\<FM\!0B\-D"V@&#8203;9QJA%&D\*S?5NOE
M\*AD&^&\<\->\\9\]J17\!"&#8203;:1I;\\ON\\N72\*4\.0W\*\,G99\[@&#8203;Q^;L^G3G^6\-JN78A%Y\[D
MY"^\,V\.8\]KHU9;AW\*2\<$&6\\\\N\_\!\-'&WF\`AW;V"\*\-K2'J@&#8203;S\!1?F\+FOP;$RZ/&#8203;:8
M\.^\(I>N\*HH6U=/QDW\,=&K$"W1\`H\<E2H\)Q14\!\.&E\]&#8203;:\+H\<\*A\,V4&HG0>>;NMK1\<
ML2^\+?\*$BE==A5PE\+\-B?^X?QQ;\\0\.$SJ\!T\-EA=\(Z;I%\_N^7N&#8203;:S\+\+;NU\]9\)L0\]
M%Y=;K>T\\Q%0^\-FDK\*D\.\]\!??\\&SSB%1\.;26K7O4$$IV0XQNR'%'VXB=BRQBX"
M=L'8ITC\\'O^\+'%XX%\!XP\)L2=\)&#8203;:S3A;1&#8203;:>H\)GJ2L\,^24D@&#8203;P2DF8\*Q%'9\]WGG\`
MEQ\(GA\-\.S\.''I\[%CC\#\)XLURWG\<@&#8203;'S\`DU4RY5CN?HG@&#8203;MS9642;\)@&#8203;4\[\.\)S9YZ\#T
\)\`6"B\(\)NS\`P\`\`
\`
end

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, May 29\, 2003 at 10&#8203;:19&#8203;:27PM \-0000\, Daniel Yacob wrote&#8203;:

> I'm using Perl 5\.8\.0 on a Redhat 8 Linux&#8203;:

The usual answer to that is "try running with LANG=C" \(RedHat defaults
to utf8 locales which are causing trouble\)
But your script doesn't look like that is the problem\.
Annoyingly I can't replicate your described output on Debian with 5\.8\.0\.

I see 2\, \(the "right" answer\, if I understand things correctly\) not 6
\(what you see\, and buggy\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From perl@geez.org

thanks for the test\, I set LC\_ALL=C and LANG=C and still get
the 6 value\.  With these settings the error message appears
four times&#8203;: "Wide character in print at \./split\-utf\-break\.pl"

So it actually worked better with a UTF8 locale \(am\_ET\.UTF\-8\)\.
Is there any additional info I can send you?

thanks\,

/Daniel

**Comment by p5pRT** [comment]
### From @andk

Nicholas Clark \<nick@&#8203;ccl4\.org> writes&#8203;:

> Annoyingly I can't replicate your described output on Debian with 5\.8\.0\.

I can and I'm also running Debian\.

> I see 2\, \(the "right" answer\, if I understand things correctly\) not 6
> \(what you see\, and buggy\)

The "6" was coming with bleadperl 18530 and is also what maint gives\.

\-\- 
andreas


Summary of my perl5 \(revision 5\.0 version 9 subversion 0 patch 18374\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.18\-k7\, archname=i686\-linux\-64int
&nbsp;    uname='linux k75 2\.4\.18\-k7 \#1 sun apr 14 13&#8203;:19&#8203;:11 est 2002 i686 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530 \-Dinstallusrbinperl=n \-Uversiononly \-Doptimize=\-g \-des \-Duse64bitint \-Dusedevel'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.2\.3 20030415 \(Debian prerelease\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldb \-ldl \-lm \-lc \-lcrypt \-lutil \-lrt
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil \-lrt
&nbsp;    libc=/lib/libc\-2\.3\.1\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.1'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: DEBUGGING USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;  Locally applied patches&#8203;:
&nbsp;        DEVEL18374
&nbsp;  Built under linux
&nbsp;  Compiled at May 30 2003 07&#8203;:05&#8203;:58
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530/lib/5\.9\.0/i686\-linux\-64int
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530/lib/5\.9\.0
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530/lib/site\_perl/5\.9\.0/i686\-linux\-64int
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530/lib/site\_perl/5\.9\.0
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/pKfBBXB/perl\-5\.8\.0@&#8203;18530/lib/site\_perl

**Comment by p5pRT** [comment]
### From @doy

I can't reproduce this\, even using the system perl on Debian \(5\.12\.4\)\.
Can anyone else reproduce it?

\-doy

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Jul 06 13&#8203;:25&#8203;:00 2012\, doy wrote&#8203;:
> I can't reproduce this\, even using the system perl on Debian \(5\.12\.4\)\.
> Can anyone else reproduce it?

I can’t\, on 5\.8\.1\, 5\.10\.1 and 5\.17\.2\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri Jul 06 13&#8203;:25&#8203;:00 2012\, doy wrote&#8203;:
> I can't reproduce this\, even using the system perl on Debian \(5\.12\.4\)\.
> Can anyone else reproduce it?

I can’t\, on 5\.8\.1\, 5\.10\.1 and 5\.17\.2\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @doy

Closing this\, since we can't reproduce it\. If someone is able to
reproduce it\, feel free to reopen this ticket\.

\-doy

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Closing this\, since we can't reproduce it\. If someone is able to
reproduce it\, feel free to reopen this ticket\.

\-doy

**Comment by p5pRT** [comment]
@doy - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Jul 06\, 2012 at 01&#8203;:54&#8203;:53PM \-0700\, Jesse Luehrs via RT wrote&#8203;:
> Closing this\, since we can't reproduce it\. If someone is able to
> reproduce it\, feel free to reopen this ticket\.

The test case in the ticket fails on the revision that Andreas mentions in
a comment in the ticket \(bleadperl 18530 \- ie 7e8c5daceba7cb18\)

Adapting the test case to die if the two values are not equal permits
bisecting\, which finds that this commit fixed it&#8203;:

d69d2d9f0d14e0c849a4b59d442938c401a7f281 is the first bad commit
commit d69d2d9f0d14e0c849a4b59d442938c401a7f281
Author&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi>
Date&#8203;:   Fri May 30 05&#8203;:47&#8203;:15 2003 \+0000

&nbsp;    Fix for "\#22375 'split'/'index' problem for utf8"\.

&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;19640

:100644 100644 d82e354341db1415bc03834f7cf84763568a16b8 310ba50465ec1a2c866438208805f6bcf626227a M      sv\.c
:040000 040000 6b5b18db188316b074b9cafdbe453ee41c29ca94 f76a7e8cef2fa3fbc79f3d2fd79ea65f83977269 M      t
bisect run success
That took 1550 seconds


The actual fix is tiny&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index d82e354..310ba50 100644
--- a/sv.c
+++ b/sv.c
@@ -5952,8 +5952,6 @@ Perl_sv_pos_b2u(pTHX_ register SV* sv, I32* offsetp)
                        }
 
                        cache[0] -= ubackw;
-
-                       return;
                    }
                }
            }




```
</details>
With some more bisecting, it turns out that the bug was introduced *at*
the commit that Andreas mentioned in the ticket&#8203;:

7e8c5daceba7cb185532328a3b67d4ca7ba4811b is the first bad commit
commit 7e8c5daceba7cb185532328a3b67d4ca7ba4811b
Author&#8203;: Hugo van der Sanden \<hv@&#8203;crypt\.org>
Date&#8203;:   Tue Jan 21 01&#8203;:37&#8203;:03 2003 \+0000

&nbsp;    integrate \(by hand\) \#18353 and \#18359 from maint\-5\.8&#8203;:
&nbsp;    Introduce a cache for UTF\-8 data&#8203;: length and byte\<\->char offset
&nbsp;    mapping are stored in a new type of magic\.  Speeds up length\(\)\,
&nbsp;    substr\(\)\, index\(\)\, rindex\(\)\, pos\(\)\, and some parts of s///\.
&nbsp;    
&nbsp;    The speedup varies a lot \(on the usual suspects&#8203;: what is the
&nbsp;    access pattern of the data\, compiler\, CPU\)\, but should be at
&nbsp;    least one order of magnitude\, and getting to the same magnitude
&nbsp;    as byte string speeds\, and in some cases  \(length on unchanged data\)
&nbsp;    even reaching the byte string speed\.  On the other hand\, in some
&nbsp;    cases \(index\) the byte speed is still faster by a factor of five
&nbsp;    or so\, but the bottleneck there does not seem to be any more
&nbsp;    the byte\<\->char offset mapping \(instead\, the fbm\_instr\(\) speed\)\.
&nbsp;    
&nbsp;    There is one cache slot for the length\, and only two for the
&nbsp;    byte\<\->char offset mapping \(the first one for the start\->offset\,
&nbsp;    and the second for the offset\->offset\+length\, when talking
&nbsp;    in substr\(\) terms\)\.
&nbsp;    
&nbsp;    Code this hairy is bound to have hairy trolls hiding under it\.
&nbsp;    \[\.\.\.\]
&nbsp;    A small tweak on top of \#18353&#8203;: don't display mg\_len bytes of
&nbsp;    mg\_ptr for PERL\_MAGIC\_utf8 because that's not what's there\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;18530


Nicholas Clark
